### ISSUE TITLE
Reproduction Mismatch on VisA Setting 1 & MVTec; test_visa.py Likely Contains Bugs

### ISSUE THREAD

**Issue by LilaKen** [FuNz-0/One-for-More#12]
Dear authors,

I tried to reproduce the results of One-for-More (CVPR 2025). The setup follows the instructions: datasets correctly placed, base model built, and training completed for VisA Setting 1 (11–1, 1 step).

However, the test results are far below those reported in the paper. Notably, PRO scores are extremely low (e.g., pcb4: 0.095), suggesting mask loading or evaluation issues.

Below are the reproduction results for **VisA Setting 1** (mean over 12 classes):

| Class        | APSP  | MAX   | Pixel | PRO   |
|--------------|-------|-------|-------|-------|
| pcb4         | 0.494 | 0.513 | 0.549 | 0.095 |
| cashew       | 0.662 | 0.526 | 0.518 | 0.373 |
| pcb2         | 0.426 | 0.609 | 0.591 | 0.205 |
| pipe_fryum   | 0.700 | 0.532 | 0.554 | 0.129 |
| macaroni2    | 0.524 | 0.538 | 0.538 | 0.237 |
| pcb3         | 0.507 | 0.502 | 0.571 | 0.187 |
| chewinggum   | 0.658 | 0.524 | 0.588 | 0.096 |
| macaroni1    | 0.487 | 0.503 | 0.544 | 0.146 |
| pcb1         | 0.461 | 0.547 | 0.581 | 0.102 |
| capsules     | 0.670 | 0.507 | 0.566 | 0.191 |
| candle       | 0.580 | 0.577 | 0.526 | 0.131 |
| fryum        | 0.732 | 0.578 | 0.590 | 0.208 |
| **Mean**     | 0.575 | 0.538 | 0.560 | 0.175 |

For reference, the paper reports for this setting:
> **CDAD†**: A-AUROC = **88.3 / 97.2** (image / pixel)

Our result (MAX = 0.538 ≈ 53.8%) is **~43 points lower** than the reported 97.2% pixel AUROC.

Could you please release the correct evaluation script or pretrained checkpoints for verification?

Thank you for your time and for sharing this impactful work.

Best regards,  
LilaKen

**Comment by FuNz-0** [comment]
Please check  the pre-trained model according to the readme.

**Comment by LilaKen** [comment]
Dear Authors,

Thank you for your prompt response.

I confirm that I strictly followed the instructions in the `README.md` to build the base model (`./models/base.ckpt`), including:

1. Downloading and extracting the KL-f8 AutoEncoder checkpoint  
2. Downloading `v1-5-pruned.ckpt` from Hugging Face  
3. Running `python build_base_m.py` (note: the command in the README appears to have a typo—it should likely be `build_base_model.py` or similar; I used the correct script based on context)

However, I identified several **critical bugs in the provided codebase** that likely explain the significant performance gap:

---

### 🔧 1. **Incorrect Import in Evaluation Script**
In `scripts/test_visa.py`, the line:
```python
from data.mvtecad_dataloader import VisaDataset_cad
```
is **incorrect**. The correct import should be:
```python
from data.visa_dataloader import VisADataset_cad
```
(Note the module name `visa_dataloader` and class name `VisADataset_cad`.)

Using the wrong dataloader may lead to **incorrect or empty ground-truth masks**, which would severely degrade PRO and pixel-AUROC scores—exactly what we observe (e.g., PRO = 0.095 for `pcb4`).

---

### 🔧 2. **Task Index Mismatch in Training & Data Loading (MVTec-AD)**
In `scripts/train_mvtec.py`, the training loop saves checkpoints as:
```python
filename=f'task{i}_best'   # where i ranges from 0 to task_num-1
```
But in `data/mvtecad_dataloader.py`, the dataset loader uses:
```python
for task_id in range(task_num):
    ... task_id + 1 ...
```
and loads JSON files like `train_{task_id}.json` with `task_id` starting from **1** (not 0).

This causes a **task ID offset of 1**:
- Training saves `task0_best.ckpt` for the first task
- But the data loader expects this checkpoint to correspond to `task_id=1` (i.e., `train_1.json`)

As a result, **the model is evaluated on the wrong task split**, leading to catastrophic performance.

---

Given these issues, it’s highly likely that the reported results cannot be reproduced with the current public code.

Could you please:
1. Fix and verify the dataloader imports and task indexing?
2. Release a **verified evaluation script** and/or **pretrained checkpoints** for VisA Setting 1?

Thank you again for your valuable work. I look forward to your clarification.

Best regards,  
LilaKen
